apiserver/common/leadership.go

```diff
@@ -0,0 +1,113 @@
+)
+
+// API exposes leadership pinning and unpinning functionality for remote use.
+type LeadershipAPI interface {
+	PinLeadership(params params.PinLeadershipBulkParams) (params.ErrorResults, error)
+	UnpinLeadership(params params.PinLeadershipBulkParams) (params.ErrorResults, error)
+}
+
+// NewFacade creates and returns a new leadership API.
+// This signature is suitable for facade registration.
+func NewLeadershipFacade(ctx facade.Context) (LeadershipAPI, error) {
+	st := ctx.State()
+	model, err := st.Model()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	pinner, err := ctx.LeadershipPinner(model.UUID())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return NewLeadershipAPI(model.ModelTag(), pinner, ctx.Auth())
+}
+
+// NewAPI creates and returns a new leadership API from the input tag,
+// Pinner implementation and facade Authorizer.
+func NewLeadershipAPI(
+	modelTag names.ModelTag, pinner leadership.Pinner, authorizer facade.Authorizer,
+) (LeadershipAPI, error) {
+	return &leadershipAPI{
+		modelTag:   modelTag,
+		pinner:     pinner,
+		authorizer: authorizer,
+	}, nil
+}
+
+type leadershipAPI struct {
+	modelTag   names.ModelTag
+	pinner     leadership.Pinner
+	authorizer facade.Authorizer
+}
+
+// PinLeadership (API) pins the leadership for applications indicated in the
+// input arguments.
+func (a *leadershipAPI) PinLeadership(param params.PinLeadershipBulkParams) (params.ErrorResults, error) {
+	entity := a.authorizer.GetAuthTag()
+	if err := a.checkAccess(entity, permission.WriteAccess); err != nil {
+		return params.ErrorResults{}, err
+	}
+
+	return a.pinOps(a.pinner.PinLeadership, param, entity), nil
+}
+
+// PinLeadership (API) unpins the leadership for applications indicated in the
+// input arguments.
+func (a *leadershipAPI) UnpinLeadership(param params.PinLeadershipBulkParams) (params.ErrorResults, error) {
+	entity := a.authorizer.GetAuthTag()
+	if err := a.checkAccess(entity, permission.WriteAccess); err != nil {
+		return params.ErrorResults{}, err
+	}
+
+	return a.pinOps(a.pinner.UnpinLeadership, param, entity), nil
+}
+
+// TODO (manadart 2018-10-19): Querying for pins,
+// which will use permission.WriteAccess
+
+func (a *leadershipAPI) pinOps(
+	op func(string, names.Tag) error, param params.PinLeadershipBulkParams, entity names.Tag,
+) params.ErrorResults {
+	results := make([]params.ErrorResult, len(param.Params))
+	for i, p := range param.Params {
+		appTag, err := names.ParseApplicationTag(p.ApplicationTag)
+		if err != nil {
+			results[i] = params.ErrorResult{Error: ServerError(err)}
+			continue
+		}
+		if err = op(appTag.Id(), entity); err != nil {
+			results[i] = params.ErrorResult{Error: ServerError(err)}
+		}
+	}
+	return params.ErrorResults{Results: results}
+}
+
+func (a *leadershipAPI) checkAccess(entity names.Tag, access permission.Access) error {
+	switch entity.Kind() {
+	case names.MachineTagKind, names.ApplicationTagKind, names.ModelTagKind:
+		return nil
```

Comment:
this would seem to say that any Machine has access to pin any application.
I don't understand what entity would be coming in as an ApplicationTag or a ModelTag. I'm guessing this might be vestigial from when we were passing in the tags?

I could live with having Machines having slightly overbroad permissions, but I would like us to think about it a bit.
We generally do try to restrict permissions to "What you need to operate and no more", as just sort of a defense-in-depth against someone getting access to a machine and then not being able to corrupt/deploy/escalate their privileges. (hacking Apache on a juju deployed machine shouldn't give you free reign on the rest of the model.)

I don't think this one is super sensitive, but just a general policy of restricting access where we can.
At least we should think about how difficult it would be if we wanted to limit the pin/unpin of applications to machine agents that have those applications deployed on them.